Assets/Player/CameraControls.gd

```diff
@@ -0,0 +1,114 @@
+extends Node
+
+const ZOOM_IN_LIMIT = 5
+const ZOOM_OUT_LIMIT = 35
+const ZOOM_VALUE = 5
+const RAY_LENGTH = 1000
+
+const MOVE_SPEED = 10
+const MOVE_FASTER_MULT = 3
+
+export(NodePath) var origin_path : NodePath
+export(NodePath) var camera_path : NodePath
+export(NodePath) var rotation_y_path : NodePath
+
+onready var _origin : Spatial = get_node(origin_path)
+onready var _camera : Camera = get_node(camera_path)
+onready var _rotation_y : Spatial = get_node(rotation_y_path)
+onready var _viewport := get_viewport()
+onready var _viewport_size := _viewport.size
+onready var _viewport_aspect := _viewport_size.aspect()
+
+var _basis : Basis
+var _move_drag_start : Vector2
+var enabled : bool = true setget set_enabled, get_enabled
+
+func _ready():
+	if _viewport.connect("size_changed", self, "_on_viewport_size_changed") != OK:
+		push_error("Failed To Connect Viewport")
+	_basis = _get_basis()
+
+func _process(delta):
+	_move(delta)
+	_move_drag()
+
+
+func _input(event):
+	if event.is_action_pressed("rotate_left"):
+		_rotate(-PI/2)
+	
+	elif event.is_action_pressed("rotate_right"):
+		_rotate(PI/2)
+	
+	elif event.is_action_pressed("zoom_in"):
+		if _camera.size > ZOOM_IN_LIMIT:
+			_zoom(-ZOOM_VALUE)
+	
+	elif event.is_action_pressed("zoom_out"):
+		if _camera.size < ZOOM_OUT_LIMIT:
+			_zoom(ZOOM_VALUE)
+
+
+func _move(delta : float):
+	var movement_scale: float = delta * MOVE_SPEED
+	if Input.is_action_pressed("move_faster"):
+		movement_scale *= MOVE_FASTER_MULT
+	
+	var x = Input.get_action_strength("move_right") - Input.get_action_strength("move_left")
+	var y = Input.get_action_strength("move_down") - Input.get_action_strength("move_up")
+	var movement_velocity = Vector3(x, 0, y * _viewport_aspect)
+	_origin.translate(_basis.xform(movement_velocity) * movement_scale)
+
+
+func _move_drag():
+	if Input.is_action_pressed("move_drag"):
+		var m_pos = _viewport.get_mouse_position()
+		if Input.is_action_just_pressed("move_drag"):
+			_move_drag_start = m_pos
+		else:
+			var drag_dir = _move_drag_start - m_pos
+			drag_dir /= _viewport_size / _camera.size
```

Comment:
Division is an expensive operation and should be minimized. Do this instead:

    drag_dir *= _camera.size / _viewport_size